ethers-middleware/src/multicall.rs

```diff
@@ -0,0 +1,386 @@
+    pub async fn run(mut self) {
+        let mut multicall: Multicall<M> =
+            Multicall::new(self.inner, self.multicall_address).await.unwrap();
+
+        loop {
+            let mut requests = Vec::new();
+
+            // wait for the first request
+            match self.rx.recv().await {
+                Some(request) => requests.push(request),
+                None => break,
+            }
+
+            // attempt to batch more requests, up to the max batch size
+            while requests.len() < self.max_batch_size {
+                match self.rx.try_recv() {
+                    Ok(request) => requests.push(request),
+
+                    // For both errors (Disconnected and Empty), the correct action
+                    // is to process the items.  If the error was Disconnected, on
+                    // the next iteration rx.recv().await will be None and we'll
+                    // break from the outer loop anyway.
+                    Err(_) => break,
+                }
+            }
+
+            let (calls, callbacks): (Vec<_>, Vec<_>) = requests.into_iter().unzip();
+
+            multicall.clear_calls();
+            for mut call in calls.into_iter() {
+                // use `to: None` as sentinel for system calls to get block number, etc
+                if call.tx.to().is_none() {
+                    call.tx.set_to(multicall.contract.address());
+                    // do not allow reverts for system calls
+                    multicall.add_call(call, false);
+                } else {
+                    // allow reverts for user calls
+                    multicall.add_call(call, true);
+                }
+            }
+            let results = multicall.call_raw().await;
+
+            let responses = match results {
+                Ok(results) => results
+                    .into_iter()
+                    .map(|result| {
+                        result.map_err(|e| {
+                            Arc::new(MulticallError::ContractError(ContractError::Revert(e)))
+                        })
+                    })
+                    .collect(),
+                Err(e) => vec![Err(Arc::new(e)); callbacks.len()],
+            };
+
+            for (callback, response) in callbacks.into_iter().zip(responses) {
+                // ignore errors, the receiver may have dropped
+                let _ = callback.send(response);
+            }
+        }
+    }
+}
+
+impl<M> MulticallMiddleware<M>
+where
+    M: Middleware,
+{
+    /// Instantiates the multicall middleware to recognize the given `match_abis`
+    pub fn new(
+        inner: M,
+        match_abis: Vec<Abi>,
+        max_batch_size: usize,
+        multicall_address: Option<Address>,
+    ) -> (Self, MulticallProcessor<M>) {
+        if max_batch_size < 2 {
+            panic!("batches must be at least 2 calls to justify the overhead of multicall");
+        }
+
+        let (tx, rx) = mpsc::channel(max_batch_size);
```

Comment:
should this channel be unbounded?

Reply:
I don't have a great intuition for this but I figured bounded was safer for memory growth?